### PR TITLE
Add BBE for jsonpath expressions

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Add BBE for jsonpath expressions

### CODE REVIEW

**Comment by sm1990** [comment]
LGTM

File: examples/jsonpath-expressions/jsonpath_expressions.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# JsonPath expressions
+
+JSON Path expressions provide a straightforward method for navigating the structure of JSON data
```
Comment (anupama-pathirage) [inline examples/jsonpath-expressions/jsonpath_expressions.md:3]:
```suggestion
JSON Path expressions provide a straightforward method for navigating the structure of JSON data.
```

File: examples/jsonpath-expressions/jsonpath_expressions.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+::: code jsonpath_expressions.bal :::
+
+::: out jsonpath_expressions.out :::
+
+## Related links
+- [Json type](/learn/by-example/json-type/)
```
Comment (anupama-pathirage) [inline examples/jsonpath-expressions/jsonpath_expressions.md:10]:
```suggestion
- [JSON type](/learn/by-example/json-type/)
```

File: examples/jsonpath-expressions/jsonpath_expressions.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+- [Access Json elements](/learn/by-example/access-json-elements/)
```
Comment (anupama-pathirage) [inline examples/jsonpath-expressions/jsonpath_expressions.md:11]:
```suggestion
- [Access JSON elements](/learn/by-example/access-json-elements/)
```

**Review by anupama-pathirage — CHANGES_REQUESTED** [review]

File: examples/jsonpath-expressions/jsonpath_expressions.bal [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        year: 1949
+    }
+];
+
+public function main() returns error? {
+    // JSON path expression to get the list of titles in the books array.
+    json titles = check jsondata:read(books, `$..title`);
+    io:println(titles);
+
+    // JSON path expression to get the list of publised years for the
```
Comment (anupama-pathirage) [inline examples/jsonpath-expressions/jsonpath_expressions.bal:30]:
```suggestion
    // JSON path expression to get the list of published years for the 
```

File: examples/jsonpath-expressions/jsonpath_expressions.bal [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    }
+];
+
+public function main() returns error? {
+    // JSON path expression to get the list of titles in the books array.
+    json titles = check jsondata:read(books, `$..title`);
+    io:println(titles);
+
+    // JSON path expression to get the list of publised years for the 
+    // books that have price value more than 80.
```
Comment (anupama-pathirage) [inline examples/jsonpath-expressions/jsonpath_expressions.bal:31]:
```suggestion
    // books that have a price value of more than 80.
```

File: examples/index.json [hunk 6]
```diff
@@ -4494,6 +4494,20 @@
       }
     ]
   },
+  {
+    "title": "Json data library",
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply examples/index.json]: Updated. 72e3d309ab19c8735b01648e10cfa5552f3ccead

File: examples/index.json [hunk 7]
```diff
@@ -4494,6 +4494,20 @@
...
+    "column": 3,
+    "category": "Common libraries",
+    "samples": [
+      {
+        "name": "Jsonpath expressions",
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply examples/index.json]: Updated. 72e3d309ab19c8735b01648e10cfa5552f3ccead

File: examples/jsonpath-expressions/jsonpath_expressions.md [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# JsonPath expressions
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply examples/jsonpath-expressions/jsonpath_expressions.md]: Updated. 72e3d309ab19c8735b01648e10cfa5552f3ccead

File: examples/jsonpath-expressions/jsonpath_expressions.md (hunk 1 above)
Comment (SasinduDilshara) [reply examples/jsonpath-expressions/jsonpath_expressions.md]: Updated. 72e3d309ab19c8735b01648e10cfa5552f3ccead

[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply examples/jsonpath-expressions/jsonpath_expressions.md]: Updated. 72e3d309ab19c8735b01648e10cfa5552f3ccead

File: examples/jsonpath-expressions/jsonpath_expressions.md (hunk 2 above)
Comment (SasinduDilshara) [reply examples/jsonpath-expressions/jsonpath_expressions.md]: Updated. 72e3d309ab19c8735b01648e10cfa5552f3ccead

File: examples/jsonpath-expressions/jsonpath_expressions.md (hunk 3 above)
Comment (SasinduDilshara) [reply examples/jsonpath-expressions/jsonpath_expressions.md]: Updated. 72e3d309ab19c8735b01648e10cfa5552f3ccead

File: examples/jsonpath-expressions/jsonpath_expressions.bal (hunk 4 above)
Comment (SasinduDilshara) [reply examples/jsonpath-expressions/jsonpath_expressions.bal]: Updated. 72e3d309ab19c8735b01648e10cfa5552f3ccead

File: examples/jsonpath-expressions/jsonpath_expressions.bal (hunk 5 above)
Comment (SasinduDilshara) [reply examples/jsonpath-expressions/jsonpath_expressions.bal]: Updated. 72e3d309ab19c8735b01648e10cfa5552f3ccead

File: examples/index.json [hunk 9]
```diff
@@ -4494,6 +4494,20 @@
...
+    "column": 3,
+    "category": "Common libraries",
+    "samples": [
+      {
+        "name": "JSONPath expressions",
+        "url": "jsonpath-expressions",
+        "verifyBuild": true,
+        "verifyOutput": true,
+        "isLearnByExample": true
+      }
```
Comment (prakanth97) [inline examples/index.json:4508]: We should move this BBE under the section `JSON` in the language concepts

**Review by LakshanWeerasinghe — DISMISSED** [review]

File: examples/jsonpath-expressions/jsonpath_expressions.bal [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import ballerina/data.jsondata;
+import ballerina/io;
+
+json books = [
```
Comment (MaryamZi) [inline examples/jsonpath-expressions/jsonpath_expressions.bal:4]: Does this have to be a module-level variable?

File: examples/jsonpath-expressions/jsonpath_expressions.bal [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    // JSONPath expression to get the list of titles in the books array.
+    json titles = check jsondata:read(books, `$..title`);
+    io:println(titles);
+
+    // JSONPath expression to get the list of published years for the 
+    // books that have a price value of more than 80.
+    json years = check jsondata:read(books, `$..[?(@.price > 80)].year`);
+    io:println(years);
+
+    // JSONPath expression to get the total summation of the books.
```
Comment (MaryamZi) [inline examples/jsonpath-expressions/jsonpath_expressions.bal:35]:
```suggestion
    // JSONPath expression to get the total sum of the prices of the books.
```

File: examples/jsonpath-expressions/jsonpath_expressions.bal [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    json totalSum = check jsondata:read(books, `$..price.sum()`);
```
Comment (MaryamZi) [inline examples/jsonpath-expressions/jsonpath_expressions.bal:36]: I would just use sum.

File: examples/jsonpath-expressions/jsonpath_expressions.md [hunk 13]
```diff
@@ -0,0 +1,11 @@
+# JSONPath expressions
+
+JSONPath expressions provide a straightforward method for navigating the structure of JSON data.
```
Comment (MaryamZi) [inline examples/jsonpath-expressions/jsonpath_expressions.md:3]: Elaborate a bit to say Ballerina supports evaluating JSONPath expressions against `json` values via the `ballerina/data.jsondata` module.

File: examples/jsonpath-expressions/jsonpath_expressions.md [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
```
Comment (MaryamZi) [inline examples/jsonpath-expressions/jsonpath_expressions.md:4]: Wonder if in a separate para we should say "Alternatively, ..." and mention about Ballerina JSON access (linking to those examples), maybe saying that's strongly typed.

Comment (SasinduDilshara) [reply examples/jsonpath-expressions/jsonpath_expressions.md:7]: updated

**Comment by prakanth97** [comment]
@SasinduDilshara can you resolve the conflict?

**Comment by SasinduDilshara** [comment]
> @SasinduDilshara can you resolve the conflict?

Done

File: examples/jsonpath-expressions/jsonpath_expressions.md [hunk 15]
```diff
@@ -0,0 +1,14 @@
+# JSONPath expressions
+
+Ballerina supports JSONPath expressions against JSON data through 
+`ballerina/data.jsondata` module to provide a straightforward method 
+for navigating the structure of JSON data.
+Alternatively, Ballerina supports type based access to JSON data by language itself
```
Comment (LakshanWeerasinghe) [inline examples/jsonpath-expressions/jsonpath_expressions.md:6]:
```suggestion
Alternatively, Ballerina supports type based access to JSON data by language itself.
```

File: examples/jsonpath-expressions/jsonpath_expressions.md [hunk 16]
```diff
@@ -0,0 +1,14 @@
+# JSONPath expressions
+
+Ballerina supports JSONPath expressions against JSON data through 
+`ballerina/data.jsondata` module to provide a straightforward method 
+for navigating the structure of JSON data.
```
Comment (prakanth97) [inline examples/jsonpath-expressions/jsonpath_expressions.md:5]: We can add this in a single line. Refer other BBEs.

File: examples/jsonpath-expressions/jsonpath_expressions.md [hunk 17]
```diff
@@ -0,0 +1,14 @@
+# JSONPath expressions
+
+Ballerina supports JSONPath expressions against JSON data through 
+`ballerina/data.jsondata` module to provide a straightforward method
```
Comment (prakanth97) [inline examples/jsonpath-expressions/jsonpath_expressions.md:4]: Shall we link API docs here.

File: examples/jsonpath-expressions/jsonpath_expressions.bal [hunk 18]
```diff
@@ -0,0 +1,38 @@
+import ballerina/data.jsondata;
+import ballerina/io;
+
+public function main() returns error? {
+    json books = [
```
Comment (prakanth97) [inline examples/jsonpath-expressions/jsonpath_expressions.bal:5]: Shall we define this as module level variable?

Comment (SasinduDilshara) [reply examples/jsonpath-expressions/jsonpath_expressions.bal:5]: Update like this according to https://github.com/ballerina-platform/ballerina-distribution/pull/5287#discussion_r1557303461

**Review by sm1990 — APPROVED** [review]

**Review by anupama-pathirage — APPROVED** [review]

**Pull request merged**